File: hackerrank/problem solving/algorithms/2 - implementation/18 - climbing the leaderboard/with morfo library/climbing_the_leaderboard.go
// https://www.hackerrank.com/challenges/climbing-the-leaderboard/problem?isFullScreen=true

package main

import (
    "morfo/morfoio"
    "slices"
)

func main() {
    morfoio.SkipInputLn()
    ranked := removeDuplicates(morfoio.ReadLn[int]())

    morfoio.SkipInputLn()
    player := morfoio.ReadLn[int]()

    morfoio.Println(climbingLeaderboard(ranked, player))
}

// n: length of arr
// k: length of distinct numbers in arr
// k <= n
// T: O(n)
// S: O(1) extra space
func removeDuplicates(arr []int) []int {
    return slices.Compact(arr)
}

// n1: length of player, playerRanks
// n2: length of ranked
// T: O(n1 + n2)
// S: O(n1) extra space
func climbingLeaderboard(ranked, player []int) []int {
    playerRanks := make([]int, len(player))
    seeker := len(ranked) - 1

    for writer, score := range player {
        for seeker >= 0 && score >= ranked[seeker] {
            seeker--
        }
        playerRanks[writer] = seeker + 2
    }

    return playerRanks
}

File: hackerrank/problem solving/algorithms/2 - implementation/18 - climbing the leaderboard/with morfo library/climbing_the_leaderboard.java
// https://www.hackerrank.com/challenges/climbing-the-leaderboard/problem?isFullScreen=true
// Java 25

import morfo.io.Reader;
import morfo.io.Writer;

void main() {
    Reader.skipInputLn();
    int[] ranked = removeDuplicates(Reader.readLnInts());
    Reader.skipInputLn();
    int[] player = Reader.readLnInts();
    Writer.println(climbingLeaderboard(ranked, player));
}

// n: length of arr
// k: length of distinct numbers in arr
// k <= n
// T: O(n)
// S: O(k) extra space
int[] removeDuplicates(int[] arr) {
    return Arrays.stream(arr).distinct().toArray();
}

// n1: length of player, playerRanks
// n2: length of ranked
// T: O(n1 + n2)
// S: O(n1) extra space
int[] climbingLeaderboard(int[] ranked, int[] player) {
    int[] playerRanks = new int[player.length];
    int seeker = ranked.length - 1;

    for (int writer = 0; writer < player.length; writer++) {
        while (seeker >= 0 && player[writer] >= ranked[seeker]) seeker--;
        playerRanks[writer] = seeker + 2;
    }

    return playerRanks;
}
